File: 547-number-of-provinces/number-of-provinces.cpp
class Solution {
public:
    void dfs(int i,vector<int> &visited,vector<vector<int>> &adj){
        visited[i]=1;
        for(auto j:adj[i]){
            if(visited[j]==0){
                dfs(j,visited,adj);
            }
            
        }
    }
    int findCircleNum(vector<vector<int>>& isConnected) {
        int n=isConnected.size();
        vector<vector<int>> adj;
        for(int i=0;i<n;i++){
            vector<int> temp;
            for(int j=0;j<n;j++){
                if(i==j) continue;
                if(isConnected[i][j]==1){
                    temp.push_back(j);
                }
            }
            adj.push_back(temp);
        }
        vector<int> visited(n+1,0);
        int cnt=0;
        for(int i=0;i<n;i++){
            if(visited[i]==0){
                cnt++;
                dfs(i,visited,adj);
            }
        }
        return cnt;
    }
};

File: 547-number-of-provinces/number-of-provinces.java
class Solution {

    public void dfs(int node, int[] visited, ArrayList<ArrayList<Integer>> adj) {
        visited[node] = 1;

        for (int neighbor : adj.get(node)) {
            if (visited[neighbor] == 0) {
                dfs(neighbor, visited, adj);
            }
        }
    }

    public int findCircleNum(int[][] isConnected) {
        int n = isConnected.length;

        ArrayList<ArrayList<Integer>> adj = new ArrayList<>();

        // Build adjacency list
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
            for (int j = 0; j < n; j++) {
                if (i == j) continue;
                if (isConnected[i][j] == 1) {
                    adj.get(i).add(j);
                }
            }
        }

        int[] visited = new int[n];
        int count = 0;

        // Count connected components
        for (int i = 0; i < n; i++) {
            if (visited[i] == 0) {
                count++;
                dfs(i, visited, adj);
            }
        }

        return count;
    }
}